Add high-level subcommands for customizing live ISO images and PXE initrds

bgilbert opened the pull request:
Fully customizing the installation process can involve a lot of moving parts, including installer config files, NetworkManager key files, kernel arguments, and nested Ignition configs.  Add user-friendly `iso customize` and `pxe customize` wrappers allowing common customizations to be performed with individual command-line options.

There are a lot of options here, and I'd like feedback on whether they provide the right set of functionality and whether they're clearly named.

[This tool](https://gist.github.com/bgilbert/6480e278d4fbc1aee769f8f567358c97) is helpful for visualizing Ignition configs generated by these subcommands.

cc @dustymabe for the networking parts in particular.

jlebon commented on src/live.rs:
```diff
@@ -736,6 +736,10 @@ struct LiveIgnition {
 impl LiveIgnition {
     pub fn from_common(common: &CommonCustomizeConfig) -> Result<Self> {
         let mut conf = Self::default();
+        // Fail at runtime if the live OS coreos-installer doesn't support
```
Not sure if it's worth it, though at least in the ISO case, we could e.g. extract `coreos-installer.service` and check for `/etc/coreos/installer.d` so that we fail at `coreos-installer iso customize` time.

jlebon commented on src/live.rs:
```diff
@@ -829,6 +836,30 @@ RequiredBy={install_target}",
+                "/etc/coreos/installer.d/{:04}-{}",
+                self.installer_serial, filename
+            ),
+            &serde_yaml::to_vec(config).context("serializing installer config")?,
```
Why reserialize here? From a debugging point of view, it'd probably be better to just drop the config verbatim so e.g. comments are retained. We parsed it, so we know it's at least valid.

jlebon commented on src/live.rs:
```diff
@@ -782,6 +788,20 @@ RequiredBy=multi-user.target"#.into(),
     }
 
+    pub fn dest_device(&mut self, device: &str, overwrite: bool) -> Result<()> {
+        println!(
+            "Warning: boot media will overwrite {} without confirmation.",
+            device
+        );
+        if !overwrite {
+            bail!("specify --overwrite-dest-device to continue");
+        }
```
Any reason for this? Our stance so far has been that you better know what you're doing re. passing the device name.

bgilbert replied:
Part of that stance has been that we won't autodetect the install device for you, because it's too dangerous; you have to specify it explicitly.  But here we're creating an image that overwrites the target disk without explicit user action, so I thought an additional safeguard would be appropriate.

However, it's reasonable to argue that we're time-shifting the explicit user action rather than eliminating it, and if the user wants a timebomb we should give it to them.  I'll drop the hidden option.

jlebon commented on src/live.rs:
```diff
@@ -732,6 +732,11 @@ fn ignition_embed_area(iso: &mut IsoFs) -> Result<Region> {
 struct LiveIgnition {
     /// The Ignition config for the live system
     live: Ignition,
+    /// The Ignition config for the destination system
+    dest: Option<Ignition>,
+    /// A user-supplied Ignition config for the dest system, which might be
+    /// merged into the dest config or might become the dest config
+    deferred_dest: Option<ignition_config::Config>,
```
Would it be cleaner and easier to understand to just keep it as a `Vec<ignition_config::Config>` and then we can decide based on `dest.len()` whether to wrap or not at `into_bytes()` time?

jlebon commented on src/cmdline.rs:
```diff
@@ -409,6 +409,9 @@ pub struct CommonCustomizeConfig {
     pub overwrite_dest_device: bool,
+    /// NetworkManager keyfile for live & dest
+    #[structopt(long, number_of_values = 1, value_name = "path")]
+    pub network_keyfile: Vec<String>,
```
This makes sense to me. Wanted to mention there's https://github.com/coreos/coreos-installer/issues/545 also related to this. In the future, we could have this knob actually use that so that it takes effect at live Ignition time.

dustymabe replied:
@bgilbert Does this imply something like `--network-keyfile=./foo.nmconnection --network-keyfile=./boo.nmconnection` i.e. allowing it to be used more than once?

@jlebon, does #545 want the network to be applied in the initramfs of the live boot?

jlebon replied:
> @jlebon, does #545 want the network to be applied in the initramfs of the live boot?

Yup exactly. I've talked with @bgilbert about this and we'll try to squeeze it in as part of this.

bgilbert replied:
@dustymabe Yes, almost all of the options can be repeated.

jlebon commented on src/cmdline.rs:
```diff
@@ -399,7 +399,7 @@ pub struct ListStreamConfig {
 
 #[derive(Debug, StructOpt)]
 pub struct CommonCustomizeConfig {
-    /// Ignition config fragment for dest system
```
Looks like this belong in some previous commit.

jlebon commented on src/cmdline.rs:
```diff
@@ -418,6 +418,9 @@ pub struct CommonCustomizeConfig {
     /// NetworkManager keyfile for live & dest
     #[structopt(long, number_of_values = 1, value_name = "path")]
     pub network_keyfile: Vec<String>,
+    /// PEM cert authority bundle for Ignition
```
For consistency and clarity, should this end with "for live & dest" ? E.g. "Ignition PEM cert authority bundle for live & dest".

bgilbert replied:
That won't fit, but we can do `Ignition PEM CA bundle for live & dest`.

jlebon reviewed (commented):
A couple of comments, but this looks great overall!

dustymabe commented:
Naming is hard...

- I'm not particularly in love with the use of `dest` but I can't really think of a much better term. `final` maybe?

- Is `dest` required in `--dest-device`. There's no `--live-device`. 

- `--ignition-ca`. Applies to both live and final system by default? Should we say that somewhere?

bgilbert commented:
> I'm not particularly in love with the use of `dest` but I can't really think of a much better term. `final` maybe?

I'm not thrilled with it either; better ideas welcome.  `final` doesn't obviously relate to the installation device IMO.

> Is `dest` required in `--dest-device`. There's no `--live-device`.

It applies to the destination system just as the other `--dest-*` options do, so I added the `dest` prefix for clarity/consistency.  Also, this keeps the name consistent with the installer config file option, which is also called `dest-device`.

bgilbert commented on src/cmdline.rs:
```diff
@@ -397,9 +397,24 @@ pub struct ListStreamConfig {
+#[derive(Debug, StructOpt)]
+pub struct CommonCustomizeConfig {
+    /// Script to run before installation
+    #[structopt(long, value_name = "path")]
```
xref https://github.com/coreos/coreos-installer/pull/305#discussion_r460055902 :grin: 

Consistency seems good, yes.  I still find SCREAMING CASE less readable, and it seems unnecessary since clap wraps value names in `<>` anyway.  But if you feel strongly about it, I can convert these over.

bgilbert commented:
Rebased, refactored for recent developments, added long help.  Not yet tested.

bgilbert commented:
In particular, `pxe customize` now takes a CoreOS initramfs and prepends it to the customizations.  This is consistent with the `iso customize` semantics, and allows both subcommands to use the new `features.json` to validate that the OS image supports the customization features we're going to use.

To validate that the input is a CoreOS live initrd, we check for `/etc/coreos-live-initramfs`, which was previously an internal API between coreos-assembler and fedora-coreos-config.  This makes the existence of that file into a stable API.

I originally implemented in-place modification of the input, as `iso customize` does by default, but implementing in-place reversion to a pristine image would be tricky.  To avoid the footgun, I omitted in-place customization for now.  I left the subcommand syntax parallel with `iso customize` (`-o` is required, `-o -` for stdout) so we could add this later.

bgilbert commented:
Also of note, some options want to inform the install process but not necessarily initiate it.  For example, `--network-keyfile` shouldn't, by itself, cause coreos-installer to run inside the live system, which would happen if we unconditionally set `copy-network` in an installer config.  So, in some cases, we need to defer processing the `dest` portion of an option until all the options have been scanned.

bgilbert commented:
- Switched to reading `features.json` directly from the ISO to improve performance
  - https://github.com/coreos/coreos-assembler/pull/2617
- Renamed dest karg options for consistency with live karg options
- Added tests

jlebon commented on tests/customize.sh:
```diff
@@ -0,0 +1,304 @@
+#!/bin/bash
```
These tests look great!

jlebon reviewed (approved):
Really nice, LGTM! The split commits are helpful.

Mostly just some tiny nits. But I'd be fine getting this in as is too and fixing things up as follow-ups.

jlebon commented on tests/customize.sh:
```diff
@@ -0,0 +1,304 @@
+    "${prog}" pxe customize src-initrd -o initrd "$@"
+}
+
+qemu_common() {
```
Would've been nice to leverage `kola qemuexec` for this, but yeah... it'd require a bunch of changes. Might be worth thinking about long-term doing the enhancements necessary and switching over. E.g. I'd love to be able to do e.g. `kola qemuexec -p pxe`.

bgilbert commented on src/cmdline.rs:
```diff
@@ -484,6 +484,13 @@ pub struct CommonCustomizeConfig {
     /// shell.
     #[structopt(long, value_name = "path")]
     pub post_install: Vec<String>,
+    /// Installer config file
```
The user will need to look up (or know) the schema anyway, so "YAML" doesn't add anything here IMO.
